[IN REVIEW] Add More Hyperelastic Materials

henrij22 opened the pull request.

henrij22 commented on ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/deviatoricinvariants.hh:
```diff
@@ -0,0 +1,119 @@
+      for (const auto j : dimensionRange()) {
+        if (i == j) {
+          ddW1dLambda(i, j) = (2.0 / 9.0) * (5 * I1 - 3 * pow(lambda_[i], 2.0)) / (pow(lambda_[i], 2.0) * I3Pow1by3);
+          ddW2dLambda(i, j) =
```
Maybe here more Eigen simplifications, but i am not sure atm how to incoporate the if statement

tarun-mitruka commented on ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/deviatoricinvariants.hh:
```diff
@@ -0,0 +1,118 @@
+    for (const auto j : dimensionRange()) {
+      for (const auto i : dimensionRange())
+        if (i != j) {
+          ddW1dLambda(i, j) = (4.0 / 9.0) * (I1 - 3 * (pow(lambda_[i], 2.0) + pow(lambda_[j], 2.0))) /
```
Saving here: I tried this but it doesn't work as `term0` has contributions from the diagonal terms. From my side, we can also leave it as such.
```cpp
    const auto lambdaProd = (lambda_ * lambda_.transpose()).eval();
    const auto lambdaSq   = lam.square().matrix().eval();
    const auto ones       = Eigen::Vector<ScalarType, dim>::Ones().eval();

    ScalarType term0 = (lambdaSq.transpose() * ones).eval()(0, 0) + (ones.transpose() * lambdaSq).eval()(0, 0);

    auto term1 = ((4.0 / 9.0) * (I1 - 3 * term0) * (lambdaProd.cwiseInverse() / I3Pow1by3)).eval();
    auto term2 = ((-4.0 / 9.0) * (2.0 * I2 - 3 * term0) * (lambdaProd.cwiseInverse() / I3Pow2by3)).eval();

    ddW1dLambda.template triangularView<Eigen::StrictlyUpper>() = term1;
    ddW2dLambda.template triangularView<Eigen::StrictlyUpper>() = term2;
    ddW1dLambda.template triangularView<Eigen::StrictlyLower>() = ddW1dLambda.transpose();
    ddW2dLambda.template triangularView<Eigen::StrictlyLower>() = ddW2dLambda.transpose();
```

rath3t replied:
yes this gets to verbose.

tarun-mitruka committed:
Add More Hyperelastic Materials

fix dyade

working blatz ko

rework structure part I

interface for volumetric + vf functions

ogden ?

some work on ogden

work on materials

+ ad test
some work that didn't get us anywhere

some more ad

hä

work on interface

WIP on feature/blatzko

documentation + tests

work

fix format + merge ogden

some more documentation

ad works (kind of?)

fix other tests due to new namespace introduction
Examples + tests work now?

Python Bindings compile?

fix tangentModuli for Ogden model with deviatoric principal stretches

fix spellings

try ogden with multiple parameters, fix bib and remove comments

fix ogden again

fix pythonn

fix testhyperelasticity

works in release mode now, but don't know why

renaming and moving files

start invariant based formulation

add checkDuplicates()

add stress computation for InvariantBased and add autodiff tests

fix

add tangentModuli for general InvariantBased model

add more volumetric functions

refactor and add Arruda-Boyce model

add Gent Model

fix second derivative w.r.t principal stretches for deviatoric functions

refactor BlatzKo

documentation

documentation for Arruda-Boyce model

add AutoDiffMat and refactor tests accordingly
refactor recoverNeoHookeTest

fix recoverNeoHookeTest

start with an interface to test the results of the deviatoric part

add certain results

fix certain results

rename variables

refactor recoverNeoHookeTest again

fix invariantbased for the undeformed configuration

refactor detC calculation

fix constructors and remove ShearModulus and BulkModulus

add stretchTags

fix safeMultiply

add checkJm in Gent

documentation arruda-boyce and gent

documentation for volumetric functions

update CHANGELOG.md
fix

more documentation

fix documentation

fix

add docu + some refactoring + add mats to testmaterials.cpp test

fix spelling

add latex shortcuts to \alpha and \lambda

check in invariantbased + precompilation of deviatoric materials

improve exponent check documentation

better template constraints

hotfix

format

autodiff -> experimental + some minor docu things
typos

fix format + typos

update CHANGELOG.md and about.md

remove trailing space

Get rid of more material models, keep ogden

I hope we can later cherry pick

Code review part I

Ogden: static constexpr

delete tests

Omit later

delete cmake

omit later

license year for non-main branched files
update comments for clarity; remove unused doubleContraction function

remove "Part" concepts

fix typos

refactor for autodiff -> not working

fix?

fix?

fix after rebase

AD from energy works

fix?

doesn't work

refactor

only run time error exists

fix documentation bib

more fixes

fix?

fix detF

Update CHANGELOG

VF are no template anymore, less restrictivity on types in dev

fixes

fix comment

Using more eigen componentwise notation

fix

minor changes from PR review
remove computeDirect usage

minor refactor

refactor

remove outdated comments

minor changes from review

delete checkthrow for NH

add missing type alias

add comment

minor fixes from review

forwarding ref

fix NoVolumetricPart

add back materials

enhance hyperelastic material tests and implementations

add more autodiff tests

fix

tarun-mitruka committed:
fix after rebase

rath3t commented on ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/arrudaboyce.hh:
```diff
@@ -0,0 +1,198 @@
+
+#pragma once
+
+#include <ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/deviatoricinvariants.hh>
```
Maybe we can also create the new material library  within Ikarus for now. Just with cmake but I will have to think that through.

tarun-mitruka replied:
Do you mean to have like `#include ikarus/materials/mechanics/...` instead of `#include <ikarus/finiteelements/mechanics/materials/...`? If so, if I recollect correctly, I think our idea was that the material library, at the moment, is only applicable to solve problems in solid mechanics and similar and hence is placed under `mechanics`, because we had the notion that if at all someone implements a material model corresponding to solve thermo-mechanical or `electro-mechanical` problems, then those can be separate folders. Nevertheless, if any such refactoring is needed, it should be in a separate PR and not be mixed with this one. Can you consider this point when looking over #358 and I resolve it now?

rath3t commented on ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/arrudaboyce.hh:
```diff
@@ -0,0 +1,198 @@
+    const auto lambdaM_       = matPar_.lambdaM_;
+    const auto beta           = 1 / pow(lambdaM_, 2.0);
+
+    for (const auto p : parameterRange())
```
Couldnt that be written as
```cpp
auto dW1dLambdaDyad = dyadic(dW1dLambda,dW1dLambda);
auto lambdadW1dLambdaDyad = dyadic(lambda.cwiseInverse(),dW1dLambda);
for (const auto p : parameterRange())
  {
     const auto factor1 = mu_ * alphas_[p] * pow(beta, p);
     const  auto factor2 = pow(W1, p) * ddW1dLambda * (p + 1);
     const  auto factor3 = pow(W1, p - 1) * dW1dLambdaDyad  * p * (p + 1);
          dS += factor1 * (factor2 + factor3);
          dS.diagonal() -= dW1dLambda.array() / lambda * factor1 * pow(W1, p) * (p + 1)
  }
```

rath3t commented on ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/deviatoricinvariants.hh:
```diff
@@ -0,0 +1,118 @@
+   */
+  auto firstDerivative() const {
+    const Invariants& invariants            = Impl::invariants(lambda_);
+    auto dW1dLambda                         = FirstDerivative::Zero().eval();
```
The zeros are not needed since it gets overwritten annyway?

rath3t commented on ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/gent.hh:
```diff
@@ -0,0 +1,188 @@
+///< Structure representing material parameters for the Gent material model.
+struct GentMatParameters
+{
+  double mu; ///< Denotes the shear modulus.
```
In arruda boyce the material parameters are written with underscores. Can you check that it is consistent? I would go without underscores for such simple structs.

rath3t commented on ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/gent.hh:
```diff
@@ -0,0 +1,188 @@
+    checkJm(W1);
+
+    for (const auto k : dimensionRange())
+      dWdLambda[k] += (mu * dW1dLambda[k] * Jm) / (2.0 * (Jm - W1) + 6.0);
```
This does not need to be a += since it only adds to zero and also this can simplified without indices.

rath3t commented on ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/invariantbased.hh:
```diff
@@ -0,0 +1,256 @@
+    for (const auto p : parameterRange())
+      for (const auto i : dimensionRange())
+        for (const auto j : dimensionRange()) {
+          auto W1pm1p       = safeMultiply<ST>(pow(W1, pex[p] - 1.0), pex[p]);
```
Wouldn't it be better to refactor this function to something like `powerTimesNumber`:

```cpp
template<typename ST>
double powerTimesNumber(ST x, int n, int m) {
    const double epsilon = std::numeric_limits<double>::epsilon();
    double result = 0.0;

    if (n == 0) {
        return m;  // x^0*m = 1*m=m
    } else if (m == 0) {
        return 0;  // Anything multiplied by 0 is 0 (also potential inf in this
                   // case)
    } else {
        if (std::abs(x) < epsilon) {
            if (n < 0 and m != 0)
                return std::numeric_limits<double>::signaling_NaN();
        }
    }
    // if constexpr (Concepts::AutodiffScalar<ST>)
      //   return std::pow(x, static_cast<ST>(n)) * m;  // Compute normally when x is not 1
      //else
    return std::pow(x, n) * m;  // Compute normally when x is not 1
}
```
https://godbolt.org/z/o5EjcWKn8

The commented portion is still not sufficient to get rid of the exponent transformation i think but is better understandable to safeMultiply, or? What do you think? Maybe I also did something wrong...

rath3t reviewed (commented):
Very nice addition! Thanks for your code. See my comments

tarun-mitruka committed:
changes from PR review

tarun-mitruka committed:
fix comment

rath3t commented on ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/invariantbased.hh:
```diff
@@ -0,0 +1,258 @@
+    const double epsilon = std::numeric_limits<double>::epsilon();
+    if (std::abs(x) < epsilon and p < 0)
+      return std::numeric_limits<double>::signaling_NaN();
+    double result = std::pow(x, p) * m;
```
Why is this working with double? at some point there waws a comment that the exponents need to be float to get the correct derivatives ? also if the intermediate result is stored as double the autodiff values are lost? Is this actually working here? Or does it work sicne this might nbe untested?

tarun-mitruka replied:
I thought it works because the exponents are already transformed using [this](https://github.com/ikarus-project/ikarus/blob/f62472dfc42e879448e814c810c980b0eec274b8/ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/invariantbased.hh#L207) function.
It is tested [here](https://github.com/ikarus-project/ikarus/blob/f62472dfc42e879448e814c810c980b0eec274b8/tests/src/testmaterialswithad.cpp#L47) with AutoDiff. I can also simplify like this
```cpp
  template <typename ST>
  ST powerAndMultiply(ST x, int p, int m) const {
    if (m == 0)
      return 0; // anything multiplied with 0 is 0
    if (p == 0)
      return m; // x^0 * m = 1 * m = m
    const double epsilon = std::numeric_limits<double>::epsilon();
    if (std::abs(x) < epsilon and p < 0)
      return std::numeric_limits<double>::signaling_NaN();
    return std::pow(x, p) * m;
  }
```

rath3t replied:
> I thought it works because the exponents are already transformed using [this](https://github.com/ikarus-project/ikarus/blob/f62472dfc42e879448e814c810c980b0eec274b8/ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/invariantbased.hh#L207) function.

I think it does not make sense because in powerAndMultiply it is explicitly converted back to an integer.
If the test work then this exponent transformation is not actually needed?

rath3t replied:
In the code snippet you gave noe the autodiff stuff would work actually.
Also do we need the `if (std::abs(x) < epsilon and p < 0)` ? I added it for completeness but we can also write an assert statement there, if this is never needed.

tarun-mitruka replied:
I had the `safeMultiply` earlier because I didn't have problems with exponent itself but only with the case $0^{-p} * 0$ which yields `nan` instead of zero. I can also revert back to `safeMultiply` as for me it is also clearer on what is happening.
I also used this function `safeMultiply` in such a way that `NaN` will never be produced and so `(std::abs(x) < epsilon and p < 0)` is not necessary.

tarun-mitruka replied:
It is anyways a private function specific to this material and so I guess `safeMultiply` is also not that bad

tarun-mitruka replied:
Ah you mean [this](https://github.com/ikarus-project/ikarus/blob/f62472dfc42e879448e814c810c980b0eec274b8/ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/invariantbased.hh#L207) transformation. I do not remember the error message I got back then but I think it was needed. I have also documented some reasons in the comments there, but if you need, I have to investigate again to find the exact messages.

rath3t replied:
>  I have also documented some reasons in the comments there, but if you need, I have to investigate again to find the exact messages.

Yes I read those comments and was confused because now It is converted back to integer anyway? If it works now the these transformatino might not be needed sicne we artificially transform forth and back all the time which actuall does nothing.

rath3t replied:
If `(std::abs(x) < epsilon and p < 0)` is not needed you can convert it to an assert statement. I think the `safeMultiply` function was not so clear because the reasoning why it is actually there i indirectly given through `x^p*m` and inside the new function you can make all decisions ( If `(std::abs(x) < epsilon and p < 0)` is not needed ) exactly in integer arithmetic and you dont have to hope that the floating point limits will workout. That was my reason to suggest the change.

tarun-mitruka replied:
You mean like [this](https://github.com/ikarus-project/ikarus/pull/342/commits/ee4e82fda0c3f239e29cc8fda044c0bc0ca5884d)?

rath3t replied:
Yes exactly.

tarun-mitruka committed:
fix

tarun-mitruka committed:
fix?

rath3t commented on ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/invariantbased.hh:
```diff
@@ -0,0 +1,232 @@
+   * \brief A function that computes \f$ x^p * m \f$.
+   *
+   * \details In the undeformed configuration, all principal stretches are equal to one. In this scenario, the
+   * deviatoric invariants W1 and W2 are zero. For any positive integer n, 0^{-n} is inf. Multiplying this with 0.0
```
This case actually never occurs because you said `std::abs(x) < epsilon and p < 0)` is never triggered? I'm again a bit confused. Otherwise this is fine even though I dont really understand what inputs can actually be expected.
